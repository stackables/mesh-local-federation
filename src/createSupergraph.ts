import {
	composeServices,
	compositionHasErrors,
} from "@theguild/federation-composition";
import { ExecutionResult, parse } from "graphql";
import type { YogaServerInstance } from "graphql-yoga";
import { executorFactory } from "./executor.js";
import { OnRemoteRequestHeadersCallback, SubgraphService } from "./index.js";

export interface CreateSupergraphOptions<T = unknown> {
	subgraphs: SubgraphService[];
	localSchema: YogaServerInstance<{}, any>;
	onRemoteRequestHeaders?: OnRemoteRequestHeadersCallback<T>;
}

export async function createSupergraph(opts: CreateSupergraphOptions) {
	const subgraphs = [...opts.subgraphs];

	// insert local to list
	subgraphs.push({
		subgraphName: "local://graphql",
		endpoint: "local://graphql",
	});

	const factory = executorFactory(opts);

	// fetch all remotes
	const subgraphsWithTypes = await Promise.all(
		subgraphs.map(async (sub) => {
			const fetch = factory.getFetch(sub);

			const response = await fetch(sub.endpoint, {
				method: "POST",
				headers: {
					"content-type": "application/json",
				},
				body: JSON.stringify({
					query: `{ _service { sdl } }`,
				}),
			});

			if (!response.ok) {
				throw new Error("service sdl call failed");
			}

			const body: ExecutionResult<any> = await response.json();
			if (body.errors) {
				throw new AggregateError(body.errors);
			}

			const federationSDL = body.data._service.sdl;

			return {
				typeDefs: parse(federationSDL),
				name: sub.subgraphName,
				url: sub.endpoint,
			};
		})
	);

	const result = composeServices(subgraphsWithTypes);
	if (compositionHasErrors(result)) {
		throw new AggregateError(result.errors);
	}
	return result.supergraphSdl.replaceAll("{", " {");
}
