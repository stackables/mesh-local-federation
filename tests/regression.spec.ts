import { createYoga } from "graphql-yoga";
import { createSupergraph } from "../src/createSupergraph.js";
import { createGatewayConfig } from "../src/index.js";
import { TestHarness, createFixture } from "./fixtures/index.js";

describe("Federation", () => {
	let harness: TestHarness;

	beforeEach(async () => {
		harness = await createFixture();
	});

	afterEach(async () => {
		await harness.stop();
	});

	test("supergraph is constructed correctly", async () => {
		const response = await createSupergraph({
			subgraphs: harness.subgraphs,
			localSchema: harness.localSchema,
			onRemoteRequestHeaders: ({ name }) => {
				return {
					authorization: name,
				};
			},
		});

		expect(response).toMatchSnapshot();
	});

	test("run basic query", async () => {
		const supergraphSDL = await createSupergraph({
			subgraphs: harness.subgraphs,
			localSchema: harness.localSchema,
			onRemoteRequestHeaders: ({ name }) => {
				return {
					authorization: name,
				};
			},
		});

		const yoga = await createGatewayConfig({
			supergraphSDL,
			localSchema: harness.localSchema,
			onRemoteRequestHeaders: ({ name }) => {
				return {
					authorization: name,
				};
			},
		});

		const fetch = createYoga({
			...yoga,
			context: ({ request }) => {
				return {
					user: request.headers.get("user-id"),
				};
			},
		}).fetch;

		const response = await fetch("/graphql", {
			method: "POST",
			headers: {
				"content-type": "application/json",
				"user-id": "aarne",
			},
			body: JSON.stringify({
				query: `{ 
					users {
						id
						identity
						account {
							name
						}
					}
					me {
						id
						identity
						name
						account {
							id
							name
						}
					}
				}`,
			}),
		});
		if (!response.ok) {
			throw new Error("service sdl call failed");
		}
		const body = await response.json();
		expect(body).toMatchSnapshot();
	});
});
