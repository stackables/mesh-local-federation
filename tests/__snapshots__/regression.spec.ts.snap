// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Federation run basic query 1`] = `
{
  "data": {
    "me": {
      "account": {
        "id": "google",
        "name": "Google",
      },
      "id": "aarne",
      "identity": "google",
      "name": "Aarne Laur",
    },
    "users": [
      {
        "account": {
          "name": "Google",
        },
        "id": "aarne",
        "identity": "fake",
      },
      {
        "account": {
          "name": "Google",
        },
        "id": "robert",
        "identity": "fake",
      },
    ],
  },
}
`;

exports[`Federation supergraph is constructed correctly 1`] = `
"
schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION)
  
  
  
  
  
  
 {
  query: Query
  
  
}


  directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

  directive @join__field(
    graph: join__Graph
    requires: join__FieldSet
    provides: join__FieldSet
    type: String
    external: Boolean
    override: String
    usedOverridden: Boolean
  ) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

  directive @join__graph(name: String!, url: String!) on ENUM_VALUE

  directive @join__implements(
    graph: join__Graph!
    interface: String!
  ) repeatable on OBJECT | INTERFACE

  directive @join__type(
    graph: join__Graph!
    key: join__FieldSet
    extension: Boolean! = false
    resolvable: Boolean! = true
    isInterfaceObject: Boolean! = false
  ) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

  directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

  scalar join__FieldSet


  directive @link(
    url: String
    as: String
    for: link__Purpose
    import: [link__Import]
  ) repeatable on SCHEMA

  scalar link__Import

  enum link__Purpose  {
    """
    \`SECURITY\` features provide metadata necessary to securely resolve fields.
    """
    SECURITY

    """
    \`EXECUTION\` features provide metadata necessary for operation execution.
    """
    EXECUTION
  }







enum join__Graph  {
  ACCOUNTS @join__graph(name: "accounts", url: "http://localhost:4006/graphql") 
  LOCAL @join__graph(name: "local", url: "http://localhost/graphql") 
  USERS @join__graph(name: "users", url: "http://localhost:4005/graphql") 
}

type Query @join__type(graph: ACCOUNTS)  @join__type(graph: LOCAL)  @join__type(graph: USERS)   {
  me: User @join__field(graph: LOCAL) 
  users: [User!] @join__field(graph: USERS) 
}

type Account @join__type(graph: ACCOUNTS, key: "id")  @join__type(graph: USERS, key: "id")   {
  id: ID!
  name: String! @join__field(graph: ACCOUNTS) 
}

type User @join__type(graph: LOCAL, key: "id")  @join__type(graph: USERS, key: "id")   {
  id: ID!
  identity: String @join__field(graph: LOCAL) 
  name: String! @join__field(graph: USERS) 
  account: Account @join__field(graph: USERS) 
}
    "
`;
